File: src/day5/part1.hs
main :: IO ()
main = do
  input <- readFile "input"
  let ids = map seatid $ lines input
  putStrLn $ show $ maximum ids

seatid :: String -> Int
seatid s 
  | length s == 10 = (row (take 7 s)*8) + col (drop 7 s)
  | otherwise = -1

col :: String -> Int 
col s = bsp s [0..7]

row :: String -> Int
row s = bsp s [0..127]

bsp :: String -> [Int] -> Int
bsp [] (x:[]) = x
bsp (c:s) x = case c of
                    'F' -> bsp s $ take (length x `div` 2) x
                    'L' -> bsp s $ take (length x `div` 2) x
                    'B' -> bsp s $ drop (length x `div` 2) x
                    'R' -> bsp s $ drop (length x `div` 2) x

File: src/day5/python/part1.py
f = open("input", "r")
lines = f.readlines()

def seatid(seat):
    return (row(seat[0:8]) * 8) + col(seat[7:len(seat)])

def row(s):
    return bsp(s, range(128))

def col(s):
    return bsp(s, range(8))

def bsp(string, arr):
    head, *rest = string
    array = arr
    if rest == []:
        return array[0]
    if head == 'F' or head == 'L':
        return bsp(rest, array[0:int(len(array)/2)])
    elif head == 'B' or head == 'R':
        return bsp(rest, array[int(len(array)/2):len(array)])

print(max([seatid(line) for line in lines]))
